### ISSUE TITLE
[distributed/ttnn]: Tensor topology not updated after all_gather_async

### ISSUE THREAD

**Issue by alnah005** [tenstorrent/tt-metal#38371]
### Component / Area

distributed

### Issue Type (optional)

Bad Outputs

### Observed

After calling `ttnn.experimental.all_gather_async` on a tensor that is sharded 
across the last dimension, the tensor's topology metadata still reports 
`PlacementShard` on the same dimension, even though the actual data has been 
gathered and each device now holds the full tensor.

```python
# Tensor starts sharded on dim 3
assert input_tt.tensor_topology().placements()[0].dim == 3

# After all_gather on dim -1
output_tt = ttnn.experimental.all_gather_async(input_tt, dim=-1, ...)

# Topology still reports sharded (INCORRECT)
assert isinstance(output_tt.tensor_topology().placements()[0], ttnn.PlacementShard)
assert output_tt.tensor_topology().placements()[0].dim == 3
```

### Expected

After all_gather_async on dimension D, the tensor topology should either:

1. Report PlacementReplicate (all devices have full data), OR
2. Not report sharding on dimension D anymore, OR
3. Update to reflect the actual data distribution

This is critical because code that relies on tensor_topology().placements()
to make decisions about data distribution will get incorrect information.

### 1. Steps (exact commands)

```
import pytest
import torch
import ttnn
from models.tt_transformers.tt.ccl import TT_CCL

@pytest.mark.parametrize(
    "device_params",
    [{"trace_region_size": 50000000, "num_command_queues": 1, "fabric_config": ttnn.FabricConfig.FABRIC_1D_RING}],
    indirect=True,
)
@pytest.mark.parametrize("mesh_device", [pytest.param((1, 8), id="1x8_grid")], indirect=True)
def test_create_last_dim_sharded_tensor(mesh_device):
    """
    Test creating a tensor sharded across the last dimension.
    """

    shape = (1, 1, 32, 1024)
    
    # Create input tensor
    input1_torch = torch.randn(shape, dtype=torch.bfloat16)
    
    # Shard on last dimension using ShardTensorToMesh
    input1_tt = ttnn.from_torch(
        input1_torch,
        device=mesh_device,
        layout=ttnn.TILE_LAYOUT,
        dtype=ttnn.bfloat16,
        mesh_mapper=ttnn.ShardTensorToMesh(mesh_device, dim=-1),
    )
    
    # Verify sharding topology
    topology = input1_tt.tensor_topology()
    placements = topology.placements()
    
    # Check number of placements (should be 1 for a single sharding)
    assert len(placements) == 1, f"Expected 1 placement, got {len(placements)}"
    
    # Check that it's a shard placement on the last dimension (dim 3 for 4D tensor)
    assert isinstance(placements[0], ttnn.PlacementShard), "Expected PlacementShard"
    assert placements[0].dim in [-1, 3], f"Expected sharding on dim 3 (last dim), got dim {placements[0].dim}"
    tt_ccl = TT_CCL(mesh_device) 

    output_tt = ttnn.experimental.all_gather_async(
        input1_tt,
        dim=-1,
        multi_device_global_semaphore=tt_ccl.get_and_cycle_ag_semaphore_handles(1),
        barrier_semaphore=tt_ccl.get_and_cycle_barrier_semaphore_handle(1),
        num_links=1,
        topology=ttnn.Topology.Linear,
    )
    
    
    # Verify output tensor has same topology as input
    output_topology = output_tt.tensor_topology()
    output_placements = output_topology.placements()
    assert len(output_placements) == 1, "Output still has 1 placement"
    assert isinstance(output_placements[0], ttnn.PlacementShard), "Output after gather is still sharded"
    assert output_placements[0].dim in [-1, 3], "Output after gather is still sharded on same dimension"
    print("Test passed: Tensor sharded on last dimension and topology verified after all_gather_async.")
   
```

save into file `test_gather.py`

```
pytest test_gather.py
```

### 2. Input data / link or description

Test file: [attach the test file you created]
- Creates 1x1x32x1024 tensor
- Shards on last dimension across 1x8 mesh
- Performs all_gather on dim=-1
- Checks topology metadata

### 3. Frequency

Always

### 1. Software Versions

OS: Ubuntu 22.04
tt-metal commit: main branch `cd8863aa1c99ad13d70b143779c28478af28ef73`
Python version: 3.10.19

### 2. Hardware Details

T3K

### Is this a regression?

Unknown

### Priority

None

### Impact

This affects any code that:

- Uses tensor_topology().placements() to determine data distribution
- Makes decisions based on topology metadata (e.g., whether to insert CCL ops)
- Validates tensor compatibility in distributed operations

Workarounds: Manually track tensor distribution state instead of relying on
topology metadata for workloads that use CCL operations.

Note: This also applies for other operations (e.g. Scatter, Linear when the weight is sharded on the output channel,...etc)
